File: loom_engine/src/main/scala/org/loom/geometry/PolygonType.scala
package org.loom.geometry

/**
 * PolygonType
 * either straight line based or spline based 
 * @author brogan
 *
 */

object PolygonType {
	 val LINE_POLYGON: Int = 0
	 val SPLINE_POLYGON: Int = 1
	 val OPEN_SPLINE_POLYGON: Int = 2  // open cubic spline: no closing edge, no fill, no subdivision
	 val POINT_POLYGON: Int = 3        // discrete point: single Vector2D, rendered as a dot
	 val OVAL_POLYGON: Int = 4         // axis-aligned ellipse: Vector2D(cx,cy) + Vector2D(cx+rx, cy+ry)
}

File: loom_swift/Sources/LoomEngine/Geometry/PolygonType.swift
/// The geometric role of a Polygon2D — determines how it is rendered
/// and whether it participates in subdivision.
public enum PolygonType: Int, Codable, CaseIterable, Equatable, Sendable {

    /// Straight-edged polygon: points are connected by line segments.
    case line = 0

    /// Closed cubic Bézier spline.
    /// Points are stored in groups of 4: [anchor, controlOut, controlIn, anchor].
    /// point.count is always a multiple of 4.
    case spline = 1

    /// Open cubic Bézier spline (same encoding as .spline, not closed).
    case openSpline = 2

    /// A single point marker. points.count == 1.
    case point = 3

    /// An axis-aligned ellipse. points == [centre, Vector2D(cx+rx, cy+ry)].
    case oval = 4

    /// Bypass types pass through every subdivision pass unchanged.
    public var isBypassType: Bool {
        self == .openSpline || self == .point || self == .oval
    }
}
